### PR TITLE
Viewport canvas for efficient rendering of enormous canvases.

### PR DESCRIPTION
**Pull request by jacob314**
I expect this code will all be replaced by hummingbird code in the future so the code quality can be a little lower than it would be otherwise.
This class is used to render the inspector tree lazily rendering a canvas broken down into convenient sized chunks.

### CODE REVIEW

File: lib/ui/viewport_canvas.dart [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+/// https://caniuse.com/#feat=intersectionobserver
+class ViewportCanvas extends Object with SetStateMixin {
+  ViewportCanvas({
+    @required CanvasPaintCallback paintCallback,
+    MouseCallback onTap,
+    MouseCallback onMouseMove,
+    VoidCallback onMouseLeave,
+    String classes,
+    this.addBuffer = true,
+  })  : _paintCallback = paintCallback,
```
Comment (kenzieschmoll) [inline lib/ui/viewport_canvas.dart]: can you use `this.paramName` for some of these?

File: lib/ui/viewport_canvas.dart [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+
+  final CanvasPaintCallback _paintCallback;
+  final MouseCallback _onTap;
+  final MouseCallback _onMouseMove;
+  final VoidCallback _onMouseLeave;
+  final Map<_ChunkPosition, _CanvasChunk> _chunks = {};
+
+  static const int maxChunks = 50;
+
+  /// Intersection observer used to detect when
```
Comment (kenzieschmoll) [inline lib/ui/viewport_canvas.dart]: used to detect when...?

File: lib/ui/viewport_canvas.dart [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+  IntersectionObserver _visibilityObserver;
+
+  CoreElement get element => _element;
+  final CoreElement _element;
+
+  /// Element containing all content that scrolls within the viewport.
+  final CoreElement _content;
+
+  // Fixed sized divs placed around the the content that is currently rendered
```
Comment (kenzieschmoll) [inline lib/ui/viewport_canvas.dart]: nit: I would add a period after rendered to break this up a bit and make it more readable.

File: test/viewport_canvas_test.dart [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+      final viewportRect =
+          viewportCanvas.element.element.getBoundingClientRect();
+
+      for (var canvas in canvases) {
+        final context = canvas.context2D;
+        expect(canvas.width, equals(chunkSize * window.devicePixelRatio));
+        final imageData =
+            context.getImageData(0, 0, canvas.width, canvas.height);
+        // Verify outer margin is transparent.
+        // This ensures the logic transofrming to chunk coordinates worked
```
Comment (kenzieschmoll) [inline test/viewport_canvas_test.dart]: typo: transforming

File: lib/ui/viewport_canvas.dart (hunk 1 above)
Comment (jacob314) [reply lib/ui/viewport_canvas.dart]: If these were positional parameters I would do that but for named parameters keeping the parameters free of underscores is important.

File: lib/ui/viewport_canvas.dart [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+  Rect rect;
+
+  void markNeedsPaint() {
+    _dirty = true;
+  }
+
+  void markPainted() {
+    _dirty = false;
+    _empty = false;
+  }
```
[the remark this answers is not in the document]
Comment (jacob314) [reply lib/ui/viewport_canvas.dart]: done

File: lib/ui/viewport_canvas.dart (hunk 2 above)
Comment (jacob314) [reply lib/ui/viewport_canvas.dart]: done

File: lib/ui/viewport_canvas.dart (hunk 3 above)
Comment (jacob314) [reply lib/ui/viewport_canvas.dart]: done

File: test/viewport_canvas_test.dart (hunk 4 above)
Comment (jacob314) [reply test/viewport_canvas_test.dart]: transformed comment :)

**Pull request merged**